src/crafting/CraftingManager.php

```diff
@@ -304,4 +326,13 @@ public function matchBrewingRecipe(Item $input, Item $ingredient) : ?BrewingReci
 
 		return null;
 	}
+
+	public function matchSmithingRecipe(Item $input, Item $addition, Item $template) : ?SmithingRecipe{
```

Comment:
This function should not be necessary. The ItemStackRequest provides an ID which tells us the recipe selected.

Reply:
(unless you want to allow SmithingTransaction to auto-detect the recipe for plugins, in which case take a look at how CraftingTransaction does this)